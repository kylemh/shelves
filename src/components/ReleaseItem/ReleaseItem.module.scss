@import '../../styles/mixins.scss';
@import '../../styles/variables.scss';

$releaseItemBorderRadius: 10px;

.ReleaseItem {
  position: relative;
  vertical-align: middle;
  padding: 1rem;
  margin: 0 1rem;
  border-radius: $releaseItemBorderRadius;
  border: 1px solid $gray;
  height: 150px;
  width: 150px;
  flex-shrink: 0;

  background-color: $white;
  box-shadow: 0 3px 1px rgba(0, 0, 0, 0.25);
  transition:
    background-color $animationDuration $animationTiming,
    box-shadow $animationDuration $animationTiming;

  @include below-screen-lg {
    height: 125px;
    width: 125px;
  }

  @include below-screen-sm {
    height: 100px;
    width: 100px;
  }

  &:hover {
    cursor: pointer;
  }
}

.title {
  margin: 0.25rem 0;
  padding: 0;
}

.artist {
  margin: 0;
  padding: 0;
}

.year {
  margin: 0.25rem 0;
  padding: 0;
}

.dragIndicator {
  position: absolute;
  left: 0;
  bottom: 0;
  border-top: 1px solid $primary;
  background-color: darken($color: $white, $amount: 5);
  border-bottom-left-radius: $releaseItemBorderRadius;
  border-bottom-right-radius: $releaseItemBorderRadius;
  color: transparentize($color: $primary, $amount: 0.5);
  font-size: 0.75rem;
  padding-bottom: 0.125rem;
  letter-spacing: 2px;
  text-align: center;

  width: 100%;
}

.isDragging {
  background-color: lighten($color: $white, $amount: 100);
  box-shadow: 0 10px 2px rgba(0, 0, 0, 0.25);
  transition:
    background-color $animationDuration $animationTiming,
    box-shadow $animationDuration $animationTiming;
}
