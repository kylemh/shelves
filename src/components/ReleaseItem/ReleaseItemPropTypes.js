import PropTypes from 'prop-types';

const ReleaseItemPropTypes = {
  id: PropTypes.number.isRequired,
  primaryArtistName: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  year: PropTypes.number,
};

export default ReleaseItemPropTypes;
