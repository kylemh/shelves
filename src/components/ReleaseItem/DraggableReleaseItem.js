import React from 'react';
import PropTypes from 'prop-types';
import { Draggable } from 'react-beautiful-dnd';
import ReleaseItem from './ReleaseItem';
import ReleaseItemPropTypes from './ReleaseItemPropTypes';

export default class DraggableReleaseItem extends React.Component {
  static propTypes = {
    ...ReleaseItemPropTypes,
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    index: PropTypes.number.isRequired,
  };

  shouldComponentUpdate(nextProps) {
    if (nextProps.index === this.props.index) {
      return false;
    }

    return true;
  }

  render() {
    const { id, index, primaryArtistName, title, year } = this.props;

    return (
      <Draggable draggableId={id} index={index}>
        {(provided, snapshot) => (
          <ReleaseItem
            primaryArtistName={primaryArtistName}
            title={title}
            year={year}
            id={id}
            innerRef={provided.innerRef}
            isDragging={snapshot.isDragging}
            {...provided.draggableProps}
            {...provided.dragHandleProps}
          />
        )}
      </Draggable>
    );
  }
}
