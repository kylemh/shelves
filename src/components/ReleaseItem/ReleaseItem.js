import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import ReleaseItemPropTypes from './ReleaseItemPropTypes';
import styles from './ReleaseItem.module.scss';

export default class ReleaseItem extends React.Component {
  static propTypes = {
    ...ReleaseItemPropTypes,
    innerRef: PropTypes.func,
    isDragging: PropTypes.bool,
  };

  static defaultProps = {
    innerRef: undefined,
    isDragging: false,
    year: undefined,
  };

  render() {
    const { innerRef, isDragging, title, primaryArtistName, year, id, ...rest } = this.props;

    return (
      <article
        className={classNames(styles.ReleaseItem, { [styles.isDragging]: isDragging })}
        ref={innerRef}
        {...rest}
      >
        <div className={styles.details}>
          <h4 className={styles.title}>{title}</h4>
          <h5 className={styles.artist}>by {primaryArtistName}</h5>
          {Boolean(year) && <h5 className={styles.year}>{year}</h5>}
        </div>

        <span className={styles.dragIndicator}>●●●</span>
      </article>
    );
  }
}
