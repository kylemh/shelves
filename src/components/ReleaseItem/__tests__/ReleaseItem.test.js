import React from 'react';
import { shallow } from 'enzyme';
import ReleaseItem from '../ReleaseItem';

describe('ReleaseItem', () => {
  const requiredProps = {
    id: 2595467,
    title: 'Cosmogramma',
    primaryArtistName: 'Flying Lotus',
    year: 2010,
  };

  it('should render', () => {
    const wrapper = shallow(<ReleaseItem {...requiredProps} />);

    expect(wrapper).toMatchSnapshot();
  });
});
