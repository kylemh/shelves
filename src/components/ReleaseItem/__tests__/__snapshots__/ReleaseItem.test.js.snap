// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReleaseItem should render 1`] = `
<article
  className="ReleaseItem"
>
  <div
    className="details"
  >
    <h4
      className="title"
    >
      Cosmogramma
    </h4>
    <h5
      className="artist"
    >
      by 
      Flying Lotus
    </h5>
    <h5
      className="year"
    >
      2010
    </h5>
  </div>
  <span
    className="dragIndicator"
  >
    ●●●
  </span>
</article>
`;
